Allow the `accession` sub-command to assign multiple IDs

andkaha opened the pull request:
This PR is related to https://github.com/NBISweden/LocalEGA-SE-Deployment/issues/630 and allows a user to assign multiple accession IDs to a set of ingested files using the `sda-admin` script.

Assigning accessions to three files, giving them the IDs `MYID001`, `MYID002`, and `MYID003`:
```shell
./sda-admin accession 'MYID%03d' 1 file1 file2 file3
```

Assigning accession IDs to all files in or beneath the path `project/data`, starting at 123.  The first two accession IDs will be `PROJ00000123` and `PROJ00000124`.  The ordering in which accession IDs are assigned will depend on the ordering of the files in the RabbitMQ queue.
```shell
./sda-admin accession 'PROJ%08' 123 project/data/
```

When assigning accession IDs to multiple files, the generated IDs are outputted as shown below:
```shell
$ ./sda-admin accession 'TEST%08x' 12 tmp/
Accession ID #12: TEST0000000c
Accession ID #13: TEST0000000d
Accession ID #14: TEST0000000e
Accession ID #15: TEST0000000f
Accession ID #16: TEST00000010
Accession ID #17: TEST00000011
```

---

Additional changes:

* Make the MQ exchange and queue prefixes configurable via command-line options or environment variables, like other settings. This makes the script completely project-independent. This was requested by @dbampalikis 
* Make the path to the `sda-cli` executable configurable with a command-line option (this was only settable by an environment variable before).
* Improvements to documentation.
* Improved error reporting whenever the RabbitMQ service sends back an error response or `curl` fails for other reasons. This was requested by @dbampalikis 
* Allow the user to override the S3 access key by either setting `S3_ACCESS_KEY` in the environment or by using `--s3-access-key`.  This allows a user to do everything apart from uploading without a `s3cmd` configuration file present.  Suggested by @kostas-kou

andkaha committed:
Allow the `accession` sub-command to assign multiple IDs

Assigning accessions to three files, giving them the IDs
"MYID001", "MYID002", and "MYID003":

./sda-admin accession 'MYID%03d' 1 file1 file2 file3

Assigning accession IDs to all files in or beneath the path
"project/data", starting at 123.  The first two accession
IDs will be "PROJ00000123" and "PROJ00000124".  The ordering
in which accession IDs are assigned will depend on the
ordering of the files in the RabbitMQ queue.

./sda-admin accession 'PROJ%08' 123 project/data/

andkaha committed:
Better logic for checking arguments

andkaha committed:
Disable Shellcheck check for variable printf format

andkaha committed:
Add forgottend documentation of environment variables

andkaha committed:
Make exchange and queue prefix configurable

andkaha committed:
Add command line option to set the path to `sda-cli`

andkaha committed:
Fix error in example

andkaha committed:
Use the `usage` function, and use a pager when asking for help

andkaha committed:
Document the fact that the script does not validate accession IDs

andkaha committed:
Make curl output errors, if there are any

andkaha committed:
Make curl fail if the response is an error

This is done by passing all curl responses through jq, detecting whether
we got a JSON object and whether that object contains an `error` key.
If this is the case, terminate with an error code and display the error
message that we got.

andkaha committed:
Output the generated accession IDs

kostas-kou reviewed (approved):
It is very good. I have a question though, since the S3 config is used only for getting the user id wouldn't be easier to put directly the LS user id as env var?
I mean if you need to e.g. ingest files for another user then you cannot get the config file of that user but instead you need to check the user id then put it in a random config and then export the config file. Seems more complicated than just export the USERID directly for example

andkaha commented:
> It is very good. I have a question though, since the S3 config is used only for getting the user id wouldn't be easier to put directly the LS user id as env var? I mean if you need to e.g. ingest files for another user then you cannot get the config file of that user but instead you need to check the user id then put it in a random config and then export the config file. Seems more complicated than just export the USERID directly for example

This is a really good point. I'm leaning towards implementing it, after getting some further input from the rest of the team.

andkaha committed:
Allow user to set S3 access key

... either by setting S3_ACCESS_KEY in the environment, or by using
--s3-access-key.

This overrides the access key in the `s3cmd` configuration file.

andkaha commented:
I implemented @kostas-kou's suggestion.  The environment variable is `S3_ACCESS_KEY`, or you may use `--s3-access-key`.

nanjiangshu reviewed (commented):
The implementation looks great to me @kusalananda. However, I have a comment which might not be closely related to this issue but could be good to fix. I think the script should not assume the existence of `s3cmd.conf` to print out the help message, since when this file does not exist, it's not straightforward to get the help.

andkaha committed:
Allow S3_ACCESS_KEY to be unsit until the last minute

Fixes buggy behaviour reported by Nanjiang.

andkaha commented:
I forgot to document the existence of the `S3_ACCESS_KEY` environment variable, but I'll do that next time I touch the script.
